### ISSUE TITLE
`set_index(sorted=True)` fails for empty partitions

### ISSUE THREAD

**Issue by bnaul** [dask/dask#6721]
On Dask master:
```
import dask.dataframe as dd
import numpy as np
import pandas as pd


df = pd.DataFrame({"id": np.repeat(['a', 'b'], 10), "x": np.random.random(20)})
ddf = dd.from_pandas(df, npartitions=2)
ddf[ddf.id == "b"].set_index("id", sorted=True).compute()  # ok without sorted=True
```
gives
```
~/model/.venv/lib/python3.8/site-packages/dask/dataframe/shuffle.py in compute_and_set_divisions(df, **kwargs)
    933
    934     if (
--> 935         sorted(mins) != list(mins)
    936         or sorted(maxes) != list(maxes)
    937         or any(a > b for a, b in zip(mins, maxes))

TypeError: '<' not supported between instances of 'str' and 'float'
```
It's pretty obvious why this fails but I'm actually not sure what the best fix is...? If `nan` is a valid index value then I guess you have to explicitly check the the len of every partition first?

**Comment by TomAugspurger** [comment]
Hmm, is it a valid index value for `set_index(..., sorted=True)` though? I suppose it's valid if and only if all the NaNs are in a partition by themselves. And probably the last partition, since I think that's where they're supposed to be sorted.

So I think your suggestion to also compute the lengths is a good one. Then the check that things are sorted is something like the current version, but only considering non-empty partitions?

**Comment by rockwellw** [comment]
I also ran into this problem, and came up with a workaround (if repartitioning is an option). This should eliminate any all-NaN partitions.
`df = df.repartition(npartitions=(df.map_partitions(len) > 0).compute().sum())`

Since we don't want `set_index` to repartition, maybe it's worth checking for NaN partitions and raising a more specific/helpful error.

**Comment by jsignell** [comment]
@rockwellw feel free to open a PR with the suggested fix for the sorted check.

**Comment by jsignell** [comment]
@bnaul do you think the informative error added by #6821 is enough to address this issue?

**Comment by bnaul** [comment]
sorry @rockwellw I missed your PR before. I'm not totally sure that the new error message is quite right either for this specific case: while it's technically true that "all" of the 0 values in an empty partition are NaN, I don't know that that really clarifies to the user that the root cause is an empty partition. repartitioning is one way to get rid of the empty partitions so that `set_index` as currently implemented will work, but it also seems like the sortedness check could just be a bit smarter and we could actually fix the issue? tl;dr I would vote to keep this open, if I have a chance I'll try to make a PR for `set_index` so this works as desired (or @rockwell feel free)

**Comment by rockwellw** [comment]
@bnaul good point. On second look, I'm realizing the fix may be as easy as 
```
mins = remove_nans(mins)
maxes = remove_nans(maxes)
```

My only concern is whether there is a more subtle reason why `sorted=True` should not be valid in this case. Logically it seems reasonable, since as you pointed out, the partition will be empty when computed (as opposed to full of NaNs).

**Comment by rockwellw** [comment]
Opened a draft PR #6829

**Comment by rockwellw** [comment]
@bnaul can we resolve this based on the fix in #6829 ?

**Comment by jsignell** [comment]
I just re-ran the original issue and it no longer throws an error. I am closing this now - thanks @rockwellw!
